### PR TITLE
Rewrite based on the official collectd Go package.

### PR DESCRIPTION
**Pull request by octo**
This allows us to easily add support for *collectd's* binary network protocol in a subsequent patch.

Ping @juliusv as requested by the contribution guidelines.

### CODE REVIEW

**Comment by juliusv** [comment]
Thanks a bunch! Will give this a look in the next days!

**Comment by juliusv** [comment]
Finally getting to this now - sorry for the delay!

File: main.go [hunk 1]
```diff
@@ -214,11 +193,22 @@ func (c collectdCollector) Describe(ch chan<- *prometheus.Desc) {
 	ch <- lastPush.Desc()
 }
 
+// Write writes "vl" to the collector's channel, to be (asynchonously)
```
Comment (juliusv) [inline main.go]: asynchonously -> asynchronously

File: main.go [hunk 2]
```diff
@@ -17,26 +17,18 @@ import (
...
-	Dsnames        []string
-	Time           float64
-	Interval       float64
-	Host           string
-	Plugin         string
-	PluginInstance string `json:"plugin_instance"`
-	Type           string
-	TypeInstance   string `json:"type_instance"`
-}
+const Timeout = 2
```
Comment (juliusv) [inline main.go]: golint nit: `main.go:31:7: exported const Timeout should have comment or be unexported`

It's not really a timeout itself, but a multiplier. Maybe lower-case it and adjust the name? `timeoutFactor`?

File: main.go [hunk 3]
```diff
@@ -50,102 +42,94 @@ var (
...
+	var valueType prometheus.ValueType
+
+	switch v := vl.Values[index].(type) {
+	case api.Gauge:
+		value = float64(v)
+		valueType = prometheus.GaugeValue
+	case api.Derive:
+		value = float64(v)
+		valueType = prometheus.CounterValue
+	case api.Counter:
```
Comment (juliusv) [inline main.go]: This case is identical to the `api.Derive` one. To make that more obvious, you could do: `case api.Counter, api.Derive:`.

**Comment by juliusv** [comment]
:+1: Looks great to me besides the small comments and seems to work fine for me. Thanks for the rewrite!

/cc @brian-brazil just in case he wants to take a look as well (he wrote the collectd exporter initially).

File: main.go (hunk 2 above)
Comment (octo) [inline main.go]: PTAL, I've added an explanation for what it does and why I named it *Timeout*. I'm happy to name it something else if you're still not convinced, of course ;)

File: main.go (hunk 1 above)
Comment (octo) [inline main.go]: Done.

File: main.go (hunk 3 above)
Comment (octo) [inline main.go]: `vl.Values` is of type `[]api.Value`, an interface type. There's a subtle difference between a case with a single type vs. a case with mutliple types: in the former, `v` will be of the (one) type listed, i.e. `api.Derive` in one and `api.Counter` in the other case, while in the latter `v` will still be of type `api.Value`, causing the `float64(v)` cast to error out.

**Comment by brian-brazil** [comment]
:+1:

File: main.go (hunk 3 above)
Comment (juliusv) [inline main.go]: Ah duh, didn't see that you're doing a type switch here. Makes sense.

File: main.go [hunk 4]
```diff
@@ -17,26 +17,21 @@ import (
...
-	Host           string
-	Plugin         string
-	PluginInstance string `json:"plugin_instance"`
-	Type           string
-	TypeInstance   string `json:"type_instance"`
-}
+// Timeout specifies the number of iterations after which a metric times out,
+// i.e. becomes stale and is removed from collectdCollector.valueLists. It is
+// modeled and named after the top-level "Timeout" setting of collectd.
+const Timeout = 2
```
Comment (juliusv) [inline main.go]: Cool to keep the name and docstring, but I'd still lower-case it (only upper-case names that you need to export to other packages).

Comment (octo) [inline main.go]: Done.

**Comment by juliusv** [comment]
Yay, thanks! Merging.

**Pull request merged**